### ISSUE TITLE
Leak on terminated app

### ISSUE THREAD

**Issue by bujdy** [Baseflow/flutter-geolocator#986]
## 🐛 Bug Report

I wanted to try to listen to locations even after i terminate app so i added 
```foregroundNotificationConfig: const ForegroundNotificationConfig(
            notificationText:
            "Example app will continue to receive your location even when you aren't using it",
            notificationTitle: "Running in Background",
            enableWakeLock: true,
          )
```
to settings. But when i terminate the app i get this leak 

```
E/ActivityThread(13871): Activity com.enviroapp.enviroapp.MainActivity has leaked ServiceConnection com.baseflow.geolocator.GeolocatorPlugin$1@576aec9 that was originally bound here
E/ActivityThread(13871): android.app.ServiceConnectionLeaked: Activity com.enviroapp.enviroapp.MainActivity has leaked ServiceConnection com.baseflow.geolocator.GeolocatorPlugin$1@576aec9 that was originally bound here
E/ActivityThread(13871): 	at android.app.LoadedApk$ServiceDispatcher.<init>(LoadedApk.java:1811)
E/ActivityThread(13871): 	at android.app.LoadedApk.getServiceDispatcherCommon(LoadedApk.java:1683)
E/ActivityThread(13871): 	at android.app.LoadedApk.getServiceDispatcher(LoadedApk.java:1662)
E/ActivityThread(13871): 	at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1819)
E/ActivityThread(13871): 	at android.app.ContextImpl.bindService(ContextImpl.java:1749)
E/ActivityThread(13871): 	at android.content.ContextWrapper.bindService(ContextWrapper.java:756)
E/ActivityThread(13871): 	at com.baseflow.geolocator.GeolocatorPlugin.onAttachedToActivity(GeolocatorPlugin.java:126)
E/ActivityThread(13871): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.add(FlutterEngineConnectionRegistry.java:155)
E/ActivityThread(13871): 	at io.flutter.plugins.GeneratedPluginRegistrant.registerWith(GeneratedPluginRegistrant.java:29)
E/ActivityThread(13871): 	at java.lang.reflect.Method.invoke(Native Method)
E/ActivityThread(13871): 	at io.flutter.embedding.engine.plugins.util.GeneratedPluginRegister.registerGeneratedPlugins(GeneratedPluginRegister.java:80)
E/ActivityThread(13871): 	at io.flutter.embedding.android.FlutterActivity.configureFlutterEngine(FlutterActivity.java:1004)
E/ActivityThread(13871): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onAttach(FlutterActivityAndFragmentDelegate.java:191)
E/ActivityThread(13871): 	at io.flutter.embedding.android.FlutterActivity.onCreate(FlutterActivity.java:459)
E/ActivityThread(13871): 	at android.app.Activity.performCreate(Activity.java:8000)
E/ActivityThread(13871): 	at android.app.Activity.performCreate(Activity.java:7984)
E/ActivityThread(13871): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
E/ActivityThread(13871): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3422)
E/ActivityThread(13871): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3601)
E/ActivityThread(13871): 	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
E/ActivityThread(13871): 	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
E/ActivityThread(13871): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
E/ActivityThread(13871): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
E/ActivityThread(13871): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/ActivityThread(13871): 	at android.os.Looper.loop(Looper.java:223)
E/ActivityThread(13871): 	at android.app.ActivityThread.main(ActivityThread.java:7656)
E/ActivityThread(13871): 	at java.lang.reflect.Method.invoke(Native Method)
E/ActivityThread(13871): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/ActivityThread(13871): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
D/FlutterGeolocator(13871): Binding to location service.
D/FlutterGeolocator(13871): Destroying location service.
D/FlutterGeolocator(13871): Stop service in foreground.
```

### Expected behavior
Wanted to fetch data in terminated state
### Reproduction steps
I didnt find anything in issues..
### Configuration
Same as in example 
**Version:** ^8.2.0

**Platform:** 
- [x] :robot: Android

**Comment by Ksardias** [comment]
I add 
`     assert pluginBinding != null;
      pluginBinding
              .getActivity()
              .unbindService(
                      serviceConnection
              );`
              to function onDetachedFromActivity in file GeolocatorPlugin.java
              
              and it look like this
              `
                private void registerListeners() {
        if (pluginRegistrar != null) {
            pluginRegistrar.addActivityResultListener(this.geolocationManager);
            pluginRegistrar.addRequestPermissionsResultListener(this.permissionManager);
        } else if (pluginBinding != null) {
            pluginBinding.addActivityResultListener(this.geolocationManager);
            pluginBinding.addRequestPermissionsResultListener(this.permissionManager);
        }
    }`
    but I'am not sure, that this is correct for anather functionality

**Comment by bujdy** [comment]
This doesnt work for me i still get leaks

**Comment by bujdy** [comment]
Now i get this message also 
`Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates. Response ID: 11`

**Comment by TommasoAzz** [comment]
Same error, but to me it happens on app startup.
geolocator version ^8.2.0, minSdkVersion 29, targetSdkVersion 31, compileSdkVersion 31.
It works fine with Android 10, Android 11 (hence, not returning any error), not working in Android 12.

**Comment by Wackymax** [comment]
@mvanbeusekom I will have a look at it

**Comment by Wackymax** [comment]
Having a look at the code it seems like even though the service is starting in the foreground it is still dependent on the activity and the flutter engine associated with that activity so the service is not truly as detached as it should be. Terminating the activity this leads to the service having to be terminated with it, which is what is happening currently.

For the most part this will allow some form of background location with the system keeping the activity and the service alive on a best effort basis. This issues will mostly present itself if you have Do Not Keep Activities enabled on Android. I will have a look to see if I can implement this to be truly background enabled even when the activities are detached from the flutter engine but this may require a large amount of work on the library and potentially introduce breaking changes.

**Comment by bujdy** [comment]
Any updates? Also i would like to ask, whats the meaning of "running in background"? Is it when app is only minimized or phone is sleeping, or when i kill/terminate app?

**Comment by Wackymax** [comment]
I am not going to have time to fix it this week, but the fix would be to stop the service when the activity is detached from the engine and also unbinding the service connect.

For now background will work when the app is minimised, and if wake lock is true it will also stay alive when the phone goes to sleep. If the system comes under memory pressure Android may still choose to kill the activity which will terminate the application which will stop the foreground service.

**Comment by Wackymax** [comment]
PR to unbind the service is up

**Comment by bujdy** [comment]
So i tried it with your fixed version. After killing app, android studio shows that app is still running. 
![image](https://user-images.githubusercontent.com/62099703/157220445-37aa5651-e643-474c-a602-58617c8b24b2.png)
Logs after i killed app:
![image](https://user-images.githubusercontent.com/62099703/157220645-6bcbabfc-bef1-4d21-ab75-8af1e8e3489f.png)

**Comment by bujdy** [comment]
@Wackymax is it only happening to me?

**Comment by bujdy** [comment]
Problem is that not only it shows application is still running, i still get leaks. For example i got another package where i listen to accelerometer. Because app is still running i get these leaks every millisecond 
![image](https://user-images.githubusercontent.com/62099703/157290304-790a15a9-0aa0-4c40-a957-9aa92889d171.png)

Even though i cancel all listeners. So there must be something wrong with killing the app when i got enabled ForegroundNotificationConfig

**Comment by mvanbeusekom** [comment]
@bujdy To me it looks like in this case it is the accelerometer plugin that is still active. 

According to the log it is trying to communicate on the method channel that no longer exists. The method channel mentioned is not used by the geolocator plugin.

Try removing the accelerometer plugin to see if the problem persists. If not you know it is the accelerometer plugin causing the current issues.

If the problem remains, it would be very helpful if you could create a small reproduction example app showing the same behavior. If you can provide such an example I can put some effort into debugging it tomorrow morning.

**Comment by Wackymax** [comment]
You are right, I can see the issue does still happen on my side. I am not sure why I didn't get it testing earlier. I am having a look on my side to see if I can identify possible solutions.

**Comment by mvanbeusekom** [comment]
@Wackymax Could be the same issue this PR is trying to solve: https://github.com/Baseflow/flutter-geolocator/pull/971

I haven't merged this one because I had some remarks but the original author doesn't seem to be keen on solving them.

**Comment by mvanbeusekom** [comment]
Yes very happy!

**Comment by mvanbeusekom** [comment]
I have just released version 3.1.2 of the geolocator_android package. This contains the fix provided by @Wackymax and should resolve this issue. @bujdy could you have another try (make sure to run `flutter pub upgrade` to update to the latest version)?

**Comment by bujdy** [comment]
Good job guys, everything is working, app is being killed. Only one problem that i get this error when i kill app.
```
D/AndroidRuntime( 7373): Shutting down VM
E/AndroidRuntime( 7373): FATAL EXCEPTION: main
E/AndroidRuntime( 7373): Process: sk.bluelemons.enviroapp, PID: 7373
E/AndroidRuntime( 7373): java.lang.RuntimeException: Unable to destroy activity {sk.bluelemons.enviroapp/sk.bluelemons.enviroapp.MainActivity}: java.lang.IllegalArgumentException: Receiver not registered: null
E/AndroidRuntime( 7373): 	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5111)
E/AndroidRuntime( 7373): 	at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:5140)
E/AndroidRuntime( 7373): 	at android.app.servertransaction.DestroyActivityItem.execute(DestroyActivityItem.java:44)
E/AndroidRuntime( 7373): 	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:176)
E/AndroidRuntime( 7373): 	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
E/AndroidRuntime( 7373): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
E/AndroidRuntime( 7373): 	at android.os.Handler.dispatchMessage(Handler.java:106)
E/AndroidRuntime( 7373): 	at android.os.Looper.loop(Looper.java:223)
E/AndroidRuntime( 7373): 	at android.app.ActivityThread.main(ActivityThread.java:7656)
E/AndroidRuntime( 7373): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime( 7373): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
E/AndroidRuntime( 7373): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
E/AndroidRuntime( 7373): Caused by: java.lang.IllegalArgumentException: Receiver not registered: null
E/AndroidRuntime( 7373): 	at android.app.LoadedApk.forgetReceiverDispatcher(LoadedApk.java:1435)
E/AndroidRuntime( 7373): 	at android.app.ContextImpl.unregisterReceiver(ContextImpl.java:1642)
E/AndroidRuntime( 7373): 	at android.content.ContextWrapper.unregisterReceiver(ContextWrapper.java:715)
E/AndroidRuntime( 7373): 	at com.baseflow.geolocator.LocationServiceHandlerImpl.disposeListeners(LocationServiceHandlerImpl.java:70)
E/AndroidRuntime( 7373): 	at com.baseflow.geolocator.LocationServiceHandlerImpl.stopListening(LocationServiceHandlerImpl.java:41)
E/AndroidRuntime( 7373): 	at com.baseflow.geolocator.GeolocatorPlugin.onDetachedFromEngine(GeolocatorPlugin.java:108)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.remove(FlutterEngineConnectionRegistry.java:273)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.remove(FlutterEngineConnectionRegistry.java:283)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.removeAll(FlutterEngineConnectionRegistry.java:291)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.engine.FlutterEngineConnectionRegistry.destroy(FlutterEngineConnectionRegistry.java:121)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.engine.FlutterEngine.destroy(FlutterEngine.java:425)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.android.FlutterActivityAndFragmentDelegate.onDetach(FlutterActivityAndFragmentDelegate.java:669)
E/AndroidRuntime( 7373): 	at io.flutter.embedding.android.FlutterActivity.onDestroy(FlutterActivity.java:676)
E/AndroidRuntime( 7373): 	at android.app.Activity.performDestroy(Activity.java:8245)
E/AndroidRuntime( 7373): 	at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1344)
E/AndroidRuntime( 7373): 	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5096)
E/AndroidRuntime( 7373): 	... 11 more

```

**Comment by mvanbeusekom** [comment]
Seems like it is trying to unregister a receiver that is already `null`. Maybe we should add another null check just to be sure, what do you think @Wackymax ?

**Comment by Wackymax** [comment]
@mvanbeusekom apologies about that, completely missed it with the example app. New PR is up

**Comment by mvanbeusekom** [comment]
No worries @Wackymax it happens. I am glad you are able to help and fix it so fast.

New version is published and all should be good now, @bujdy can you confirm?

**Comment by bujdy** [comment]
I tried it with newer version and i dont get any errors in console, but app is still "running" after killing it :D 
![image](https://user-images.githubusercontent.com/62099703/157412606-4d87d19c-49d9-470a-815c-d9d185a980b6.png)

**Comment by Wackymax** [comment]
This behavior occurs in older versions of the library as well and I think that we should investigate as a separate issue. But it might just be the dev tooling in this case

**Comment by bujdy** [comment]
And i dont think its only happening in dev enviroment because my friend download this app on real device, and after few hours when he tried to turn on application, whole display was black with flutter logo in the middle. I have never dealt with this in flutter.

**Comment by bujdy** [comment]
If you dont have time (which is totally fine and i understand it) maybe i would rather have previous version, where app is killed (i wouldnt mind seeing error when app is killed anyway)

**Comment by mvanbeusekom** [comment]
@bujdy i think the exception is actually the cause of the application being terminated. Leaving the exception is not a solution.

In your case you could pin the previous version using a dependency override:

```yaml
  dependency_overrides:
    geolocator_android: 3.1.2
```

**Comment by bujdy** [comment]
Thanks, i know its not good solution, but its better than leaving app opened and having leaks :D

**Comment by Wackymax** [comment]
I've tried now with an older version of this library (before the change to a service), i've tried with an older version of my own app where I don't even use this lib, and I've tried with the latest version of my app with these changes and they all behave the same. Closing the activity does not kill the debugging process, force closing the activity kills the debugging process. I am not sure what else to do anymore.

**Comment by mvanbeusekom** [comment]
Just for my comprehension, how do you close the activity? I mean what is the proces?

Is this done programmatically or by sending the app to the background and physically closing the app (swiping it away from the running apps)?

**Comment by Wackymax** [comment]
I've enabled the "Do Not Keep Activities" in the developer options of the Android settings. Then sending the app to the background should terminate/close the activity. I am trying to look now at other libraries that have faced similar issues to see if I can figure out what is going on that way. Here is another example: https://github.com/ryanheise/audio_service/issues/225

**Comment by mvanbeusekom** [comment]
Check I will try to investigate this also. Just so I have a feeling on the problem and see if I can help out.

**Comment by Wackymax** [comment]
Thank you.

**Comment by Wackymax** [comment]
Looking at similar issues reported for other libraries this is a mix between issues with the flutter engine and also having open streams with EventChannels when the activity is closed. I don't see anything that we are doing wrong but will keep looking to see if this is something we can solve. https://github.com/flutter/flutter/issues/28651 and https://github.com/alann-maulana/flutter_beacon/issues/80 and https://githubmemory.com/repo/hemanthrajv/flutter_compass/issues/36

**Comment by Wackymax** [comment]
Only way I can see we can look at doing this is sending a complete signal to close the streams when the activity is detached but I am not sure if that is the correct approach to take here. But I'm not convinced that will solve this issue. @mvanbeusekom I think it's your call here over what the next steps would be.

**Comment by mvanbeusekom** [comment]
Goodmorning @Wackymax,

I have been looking into this issue a bit more detailed using the example app that is part of the `geolocator_android` package and found out that if I comment out lines 108 to 110 in the `StreamHandlerImpl.java` file (lines are listed below) the application terminates appropriately (there is a little delay of 1 or 2 seconds, but that is normal).

```java
    if (foregroundNotificationOptions != null) {
      foregroundLocationService.enableBackgroundMode(foregroundNotificationOptions);
    }
```

Another thing I noticed is that if I remove these lines and kill the App the debugger shows the following message in the debug window:
```
Disconnected from the target VM, address: 'localhost:65496', transport: 'socket'
```

As it noticed the App is gone from the emulator and should stop debugging. If I leave the lines active and enable background mode this message doesn't appear in the debug logs which indeed seems like there is still something active. 

So it does seem like something related to the background implementation is not quite correct and the service is not stopped once the application is terminated.

**Comment by Wackymax** [comment]
Morning!

I tried it now with the steps you outlined and I am not getting the same results. On my side the example app behaves the same regardless of whether I comment out those lines or not the debugger stays active. This is making it tricky to debug.

**Comment by mvanbeusekom** [comment]
I see I haven't been very detailed in my instructions, maybe it helps if I describe it a bit more detailed.

First thing I did was build the example app by running `flutter build appbundle --release` from the `geolocator_android/example` folder. This will allow me to open the Android project in Android Studio. 

Next I opened the `geolocator_android/example/android` folder in Android Studio (IntelliJ also behaves the same). This will allow me to run and debug the Java code of the geolocator_android plugin. From here I took the following steps:

1. Selected the emulator I want to debug on (in my case a Pixel XL with API 31);
2. Start debugging the application by clicking the bug icon in the task bar (or press Ctrl+D);
3. Once the App has started I hit the red play button to start listening to the position stream;
4. While listening to the position stream I hit the Android Overview button (the square button at the bottom of the screen);
5. Next I swipe the example App away to kill it.

After step 5 I see a clear difference between having the lines (108 to 110) commented out or active.

**Comment by Wackymax** [comment]
So what I am doing differently is running the flutter example app directly from the example folder. It builds and uses the Android code that I change in the library project (confirmed with log changes). In that case step 5 of terminating the app behaves the same on my side with the app terminating correctly in both cases. Let me try your approach and see.

**Comment by mvanbeusekom** [comment]
Something else I noticed is that if you go to "Settings" -> "Apps" -> "geolocator_example" the "Force stop" button is still active when after you close the App while having the position stream active. This indicates the process is indeed still running. 

Pressing the "Force stop" button will stop the debugging process in Android Studio.

P.S. I am debugging on the `master` branch

**Comment by Wackymax** [comment]
I managed to follow your steps exactly as outlined and the application closed successfully on my side for both scenarios. I did notice the delay between closing does appear to be semi random.

**Comment by Wackymax** [comment]
Here some images of my logs.

With foreground
<img width="832" alt="Screenshot 2022-03-10 at 12 45 45" src="https://user-images.githubusercontent.com/11558269/157646236-0b189b1c-ed16-4226-b088-0b575e33c7c1.png">

Without foreground (commented out as you did)
<img width="855" alt="Screenshot 2022-03-10 at 12 46 54" src="https://user-images.githubusercontent.com/11558269/157646276-edbbdf24-cfd0-4ac2-922d-f2c2ed3407e6.png">

In both cases I followed your steps exactly.

**Comment by mvanbeusekom** [comment]
I noticed your logs contain the message "Destroyed location service", I am not able to find this message in my code base. Could it be you have a slightly different version?

**Comment by mvanbeusekom** [comment]
Goodmorning @Wackymax, 

Did you catch my previous message? Going from the logs you supplied it looks like we are looking at different code bases, which might explain the difference in behavior.

**Comment by Wackymax** [comment]
Morning! Apologies looks like my reply on the Github mail did not pull into this conversation. I've added a log entry to confirm that the onDestroy method runs until the end. Other than that everything else is as on master. Happy to remove the log entry and rerun the test for you?

<img width="1041" alt="image" src="https://user-images.githubusercontent.com/11558269/157825001-4b0c4f91-205c-40e0-9ef8-f942db33beb3.png">

**Comment by mvanbeusekom** [comment]
That explains the difference in the logs, thanks for confirming. No need to remove it and test again that shouldn't make a difference.

**Comment by bujdy** [comment]
So other plugins got same problem?

**Comment by mvanbeusekom** [comment]
@bujdy I haven't found the problem yet. The problem is that while I can reproduce the problem, @Wackymax is not able to do so. Since I don't have any experience with foreground services I also cannot tell if this is expected behaviour or really a bug. Android's documentation is also not really clear about it unfortunately.

With my limited knowledge the code seems correct and I can confirm that the right methods (`unbind` and `stopSelf`) are being called in the correct order.

**Comment by bujdy** [comment]
So there isn´t way to fix this and kill app? Or are you still investigating this issue?

**Comment by bujdy** [comment]
Any news? @Wackymax @mvanbeusekom

**Comment by royer** [comment]
my project also has this problem, I just start my app and then terminate immediately. so nothing Geolocator method called. there have some leaked information in debug console.
 
```
D[/FlutterGeolocator]()( 5205): Creating service.
D[/FlutterGeolocator]()( 5205): Binding to location service.
E[/ActivityThread]()( 5205): Activity com.example.hunteraide.MainActivity has leaked ServiceConnection com.baseflow.geolocator.GeolocatorPlugin$1@7f87ea8 that was originally bound here
E[/ActivityThread]()( 5205): android.app.ServiceConnectionLeaked: Activity com.example.hunteraide.MainActivity has leaked ServiceConnection com.baseflow.geolocator.GeolocatorPlugin$1@7f87ea8 that was originally bound here
E[/ActivityThread]()( 5205): 	at com.baseflow.geolocator.GeolocatorPlugin.onAttachedToActivity(GeolocatorPlugin.java:126)
D[/FlutterGeolocator]()( 5205): Binding to location service.
D[/FlutterGeolocator]()( 5205): Destroying location service.
D[/FlutterGeolocator]()( 5205): Stop service in foreground.
D[/FlutterGeolocator]()( 5399): Creating service.
2
D[/FlutterGeolocator]()( 5399): Binding to location service.
D[/FlutterGeolocator]()( 5399): Destroying location service.
D[/FlutterGeolocator]()( 5399): Stopping location service.
```

and then I build a very fresh flutter project, do listening location even use foregroundNotificationConfig  in this new project, everything is ok. no leaked information output.

I also make sure the compileSdkVersion and targetSdkVersion is same in both project.

I hope these information is useful.

**Comment by royer** [comment]
I found my problem. I checked pubspec.lock , find geolocator_android is 3.1.0, and the new fresh project is 3.1.3. I delete the geolocator in pubspec.yaml, and add it back, then geolocator_android is 3.1.3. the leaked information is gone.

**Comment by bujdy** [comment]
But it still wont kill the application when you are trying to close it.

**Comment by bujdy** [comment]
@Wackymax @mvanbeusekom is there anything you tried in past days?

**Comment by Wackymax** [comment]
Hi,

I haven't since I can't reproduce it on my side.

**Comment by mvanbeusekom** [comment]
It is on my list but haven't had the time yet. Might be a couple of weeks. So if anyone wants to give it a go please feel free.

**Comment by bujdy** [comment]
Any news on this?

**Comment by mvanbeusekom** [comment]
I have tried with the latest updates in the Android package and no longer experience any issues anymore.

**Comment by Wackymax** [comment]
It's a single log entry I added to confirm the code reached the end of the destroy method. Rest of the code is the same as the master branch.

**Comment by bazrafkan** [comment]
I have same problem with last version of geolocator_android (4.1.3) and geolocator(9.0.2)

E/FlutterGeolocator(24168): Geolocator position updates stopped
E/FlutterGeolocator(24168): There is still another flutter engine connected, not stopping location service

**Comment by Wackymax** [comment]
> I have same problem with last version of geolocator_android (4.1.3) and geolocator(9.0.2), Devices: Pixel 2 and Pixel 5, I think this issue have same resone with this issue #1142.
> 
> @Wackymax Do you have any news? Thanks
> 
> E/FlutterGeolocator(24168): Geolocator position updates stopped E/FlutterGeolocator(24168): There is still another flutter engine connected, not stopping location service

I am not entirely sure what you are asking? What you are seeing in the log messages is that you have multiple Flutter engines using the location service so it's not going to shut down otherwise the other Flutter engine won't receive location events.

I am also not seeing nor am I able to reproduce the issues reported by @bujdy

**Comment by Hhanri** [comment]
I'm still facing this issue when I swipe the app (geolocator 9.0.2 with geolocator_android 4.1.9).
Also the notification and the location logo in status bar doesn't go away unless I force stop the app.

```
D/FlutterGeolocator(20710): Detaching Geolocator from activity
D/FlutterGeolocator(20710): Flutter engine disconnected. Connected engine count 1
D/FlutterGeolocator(20710): Disposing Geolocator services
E/FlutterGeolocator(20710): Geolocator position updates stopped
E/FlutterGeolocator(20710): There is still another flutter engine connected, not stopping location service
E/libEGL  (20710): call to OpenGL ES API with no current context (logged once per thread)
W/FlutterJNI(20710): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 18
```

**Comment by Hhanri** [comment]
I'm still facing this issue when I swipe the app (geolocator 9.0.2 with geolocator_android 4.1.9).
Also the notification and the location icon in status bar doesn't go away unless I force stop the app.

```
D/FlutterGeolocator(20710): Detaching Geolocator from activity
D/FlutterGeolocator(20710): Flutter engine disconnected. Connected engine count 1
D/FlutterGeolocator(20710): Disposing Geolocator services
E/FlutterGeolocator(20710): Geolocator position updates stopped
E/FlutterGeolocator(20710): There is still another flutter engine connected, not stopping location service
E/libEGL  (20710): call to OpenGL ES API with no current context (logged once per thread)
W/FlutterJNI(20710): Tried to send a platform message to Flutter, but FlutterJNI was detached from native C++. Could not send. Channel: flutter.baseflow.com/geolocator_updates_android. Response ID: 18
```

**Comment by mvanbeusekom** [comment]
@Hhanri did you configure a timeout on the future returned? 

If so you should have received a `TimeoutException` and this should also cancel/ stop listening for position updates.

**Comment by Hhanri** [comment]
@mvanbeusekom I just tried that but it doesn't seem to stop the service when the app is already killed (swiped), notification and location icon are still showing up in the status bar and there is no ```TimeoutException``` in the console.
I tried ```background_locator_2``` and ```location``` plugins, they both stop the service when the app is killed so I'm wondering why ```geolocator``` doesn't.

**Comment by Wackymax** [comment]
Based on the logs you sent it seems like it's staying connected because there are multiple flutter engines alive, almost like there is a background process running with another flutter instance.What other plugins do you have installed and active?On 15 Jun 2023, at 17:17, Hanri ***@***.***> wrote:﻿
@mvanbeusekom I just tried that but it doesn't seem to stop the service when the app is already killed (swiped), notification and location icon are still showing up in the status bar and there is no TimeoutException in the console.
I tried background_locator_2 and location plugins, they both stop the service when the app is killed so I'm wondering why geolocator doesn't.

**Comment by Hhanri** [comment]
Here are all the dependencies I'm using
```
  google_sign_in: ^6.1.0
  firebase_core: ^2.9.0
  firebase_app_check: ^0.1.1+9
  firebase_auth: ^4.4.0
  cloud_functions: ^4.1.0
  cloud_firestore: ^4.5.0
  firebase_messaging: ^14.4.0
  firebase_remote_config: ^4.0.0
  firebase_analytics: ^10.2.0
  firebase_ui_firestore: ^1.3.0
  google_maps_flutter: ^2.2.5
  flutter_facebook_auth: ^4.4.1+1
  sign_in_with_apple: ^4.3.0
  crypto: ^3.0.2

  go_router: ^6.2.0
  equatable: ^2.0.5
  easy_localization: ^3.0.1
  flutter_native_splash: ^2.2.17
  flutter_config: ^2.0.0
  intl: ^0.17.0
  shared_preferences: ^2.0.15
  package_info_plus: ^3.0.2
  streaming_shared_preferences: ^2.0.0
  url_launcher: 6.1.9
  add_2_calendar: ^2.2.3
  flutter_native_timezone: ^2.0.0
  timezone: ^0.9.0
  cached_network_image: ^3.2.3
  flutter_local_notifications: ^13.0.0
  dartz: ^0.10.1
  open_mail_app: ^0.4.5
  in_app_review: ^2.0.6
  share_plus: ^6.3.0
  geolocator: ^9.0.2
  permission_handler: ^10.3.0

  flutter_animation_progress_bar: ^2.3.1
  dotted_border: ^2.0.0+3
  font_awesome_flutter: ^10.3.0
  lottie: ^2.0.0
  flutter_rating_bar: ^4.0.1

  qr_flutter: ^4.0.0
  image_picker: ^0.8.7+4
  image_cropper: ^3.0.3
  image: ^4.0.16
```
I don't use multiple location plugins at the same time when testing, always one at a time.
```background_locator_2``` and ```location``` plugins have no issue stopping the service when the app is killed.

**Comment by Hhanri** [comment]
@Wackymax I just found out it was caused by Firebase Messaging plugin, the ```FirebaseMessaging.onBackgroundMessage``` method reinitializes the Flutter Engine. 
Therefore the ```connectedEngines``` is set to 2 and ``connectedEngines == 0``` inside ```canStopLocationService``` returns false.
Is it safe to ignore ```connectedEngines```'s value and return true in ```canStopLocationService``` ?

**Comment by Wackymax** [comment]
We can't do it in the official lib because it's needed for when geolocator is running in background plugins which is expected behavior.We would have to change the code to see which engine is currently using location services and possibly stop it when we see that engine gets destroyed.I do wonder though, if you want to location services to stop with the activity then why are you setting the foreground notification in the first place? If you do not set it then the service should behave correctly and stop when the activity is stopped.On 17 Jun 2023, at 16:23, Hanri ***@***.***> wrote:﻿
@Wackymax I just found out it was caused by Firebase Messaging plugin, the FirebaseMessaging.onBackgroundMessage method reinitializes the Flutter Engine.
Therefore the connectedEngines is set to 2 and ``connectedEngines == 0insidecanStopLocationService``` returns false.
Is it safe to ignore ```connectedEngines```'s value and return true in ```canStopLocationService``` ?

**Comment by Hhanri** [comment]
@Wackymax I need to get the device's position in background, but I still want to let the user stop the service by simply closing the app (it's not a main feature).
When I run without the foreground notification I can't retrieve the position from background and it stops by itself after a few minutes.
I forked the project and bypassed ```connectedEngines == 0```, it stopped the service as I expected but I just wanted to make sure that it was safe to do it this way.

**Comment by Wackymax** [comment]
Yeah it should be safe to do that. On 17 Jun 2023, at 16:54, Hennie Brink ***@***.***> wrote:﻿We can't do it in the official lib because it's needed for when geolocator is running in background plugins which is expected behavior.We would have to change the code to see which engine is currently using location services and possibly stop it when we see that engine gets destroyed.I do wonder though, if you want to location services to stop with the activity then why are you setting the foreground notification in the first place? If you do not set it then the service should behave correctly and stop when the activity is stopped.On 17 Jun 2023, at 16:23, Hanri ***@***.***> wrote:﻿
@Wackymax I just found out it was caused by Firebase Messaging plugin, the FirebaseMessaging.onBackgroundMessage method reinitializes the Flutter Engine.
Therefore the connectedEngines is set to 2 and ``connectedEngines == 0insidecanStopLocationService``` returns false.
Is it safe to ignore ```connectedEngines```'s value and return true in ```canStopLocationService``` ?

**Comment by burhankhanzada** [comment]
deoes 

> @Wackymax I need to get the device's position in background, but I still want to let the user stop the service by simply closing the app (it's not a main feature). When I run without the foreground notification I can't retrieve the position from background and it stops by itself after a few minutes. I forked the project and bypassed `connectedEngines == 0`, it stopped the service as I expected but I just wanted to make sure that it was safe to do it this way.

does this fix it in the latest version coz i m also facing same issue m also using firebase cloud messaging

**Comment by burhankhanzada** [comment]
also is it possible to show a button in the notification if the user doesn't want to run this service while running the app or even when the app is terminated?

**Comment by thalesluquetti** [comment]
Is there any update?

**Comment by Akshay-Kumar-Sah-Mass** [comment]
I did this and it worked

  Future<void> stopLocationTracking() async {
    _positionStream?.pause();
    await _positionStream?.cancel();
    _positionStream = null;
    await Future.delayed(const Duration(milliseconds: 500));

    print("Location tracking forcefully stopped on back button pressed");
  }
